MM-12148: Prevent emdash convert in a Code Block on ios

ami9000 opened the pull request:
#### Summary
This PR adds a fix to prevent iOS from converting double dash (--) / triple dash (---) to an emdash (–) while typing a Code Block (w/ triple backtick ``` markdown). 

The iOS smart dash behavior occurs based on the keyboard type ( ‘default’). The proposed solution here is to change keyboardType to ‘email-address’ when cursorPosition is within the bounds of a Code Block and back to ‘default’ when otherwise.

The TextInput’s onSelectionChange handles cursor movements without key strokes. The TextInput’s onChangeText handles cursor movements upon new keystrokes once the TextInput’s (state) value has finished updating.

#### Ticket Link
Ticket https://mattermost.atlassian.net/browse/MM-12148
Fixes https://github.com/mattermost/mattermost-server/issues/9432

#### Checklist
- [x] Has UI changes - keyboardType changes while typing a Code Block

#### Device Information
This PR was tested on: iPhone 7 simulator, 13.0;  iPhone 11 simulator, 13.0

#### Screenshots
![emdash](https://user-images.githubusercontent.com/20244930/73701800-91517300-469f-11ea-81c6-48e126e39f70.gif)

chuttam commented on app/components/post_textbox/post_textbox_base.js:
````diff
@@ -396,17 +396,68 @@ export default class PostTextBoxBase extends PureComponent {
-        });
+        // Workaround to avoid iOS emdash autocorrect in Code Blocks
+        if (Platform.OS === 'ios') {
+            // Matches all text between a pair of triple backticks ```, inclusive
````
This could go into a separate, testable function, called from `handlePostDraftSelectionChanged` when relevant. Perhaps more descriptively named `switchKeyboardForCodeBlocks`? This could reduce the need to rely on explanatory comments.

ami9000 replied:
I separated to ` switchKeyboardForCodeBlocks` and removed extra comments. Thanks for the rec!

chuttam commented on app/components/post_textbox/post_textbox_base.js:
````diff
@@ -396,17 +396,68 @@ export default class PostTextBoxBase extends PureComponent {
+            const regexForCodeBlock = /^```$(.*?)^```$|^```$(.*)/gms;
+
+            // Loop through this.state.value and create array of all regex match positions, start and end
+            // exec() method details: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/exec
````
Recommend deleting this comment. Adds bulk and code is self-explanatory.

chuttam commented on app/components/post_textbox/post_textbox_base.js:
```diff
@@ -396,17 +396,68 @@ export default class PostTextBoxBase extends PureComponent {
+                matches.push({
+                    start: regexForCodeBlock.lastIndex - match[0].length, // start position of a regex match
+                    end: fromHandleTextChange ? regexForCodeBlock.lastIndex : regexForCodeBlock.lastIndex + 1, // end position of a regex match
+                });
```
Could this be a `.map()` on `regexForCodeBlock.exec(this.state.value)` to build out the `matches` object array.

ami9000 replied:
I revised this to a cleaner version with `.map()` and `string.matchAll(regExp)`. However, turns out `matchAll` is only supported in node12 and up as shown in the now failed test.

`regExp.exec(string)` requires a loop to build out the object array since its output isn't actually an array with all matches so we can't run `map()` directly on it. 

I'm unsure of the importance of the node environment relative to this so I defer to you here for advice.

chuttam replied:
I see. Let's go back to what you had before (using `.exec`) and manually creating the array of cursor position ranges.

Now that we have this function in two places, it should move into a single spot (perhaps `app/utils/markdown`) and with unit tests. That way, we can safely tweak this logic of cursor position range checking (and even the regex, if necessary) while ensuring we don't change the desired/expected behaviour.

Another option that came to mind was along the lines of:
```
    const matches = value.match(regexForCodeBlock).map(match => {
      return {
        start: //start logic
        end: // end logic
      }
    });

```
but let's wrap the current functionality in tests first before changing anything.

chuttam commented on app/components/post_textbox/post_textbox_base.js:
```diff
@@ -396,17 +396,68 @@ export default class PostTextBoxBase extends PureComponent {
+                if (cursorPosition >= matches[i].start && cursorPosition <= matches[i].end) {
+                    insideCodeBlock = true;
+                    break;
+                }
```
... followed by `reduce` after the preceding `map` to determine whether the cursor position is inside any code block (i.e. within any element of `matches`)

ami9000 replied:
I think I achieved what you had in mind with `.some()`

chuttam replied:
Indeed.

chuttam commented on package-lock.json:
```diff
@@ -7065,8 +7065,7 @@
           "version": "2.1.1",
-          "bundled": true,
-          "optional": true
+          "bundled": true
```
Code changes contain no dependency updates/alterations. Do not recommend committing `package-lock.json`.

ami9000 replied:
Got it!

ami9000 commented:
@chuttam Glad to contribute (this is my first)! One of the comments in the associated Jira ticket led me down this solution path. Apple documentation for `smartDashesType` mentioned keyboard type being of importance.

I added this functionality to `edit_post` as well which covers pre-existing posts like you mentioned. Sorry I missed this – just getting familiar with the codebase!

Cleaned up the code as well based on comments – ready for re-review.

chuttam commented on app/screens/edit_post/edit_post.js:
```diff
@@ -220,7 +250,8 @@ export default class EditPost extends PureComponent {
                             disableFullscreenUI={true}
                             keyboardAppearance={getKeyboardAppearanceFromTheme(this.props.theme)}
-                            onSelectionChange={this.onPostSelectionChange}
+                            onSelectionChange={(event) => this.onPostSelectionChange(event, false)}
```
This arrow function will create a new function object every time, and since it's a prop, will trigger a re-render every time. Should move to its own function definition.

ami9000 replied:
Anonymous functions. Got it!

enahum commented:
Although this seems to do the trick i’m wondering why not implement the native solution as this seems to be a little bit overhead by running a constant regex. What is the perf impact of this solution? How is it compared to the native solution?

I don’t mind the keyboard changing types when entering a code block but if this can be simplified that would be awesome.

Another thing to note is that I may want to send a message line `somecommand —flag` and not have the emdash.

ami9000 commented:
@chuttam Thanks! I’ve revised and moved cursor/keyboard logic to its own utils function and added unit tests.

@enahum I’ve tested the regex solution with the 16,383 character input limit (on iPhone 7 device) without noticing performance effects while typing. This could use a QA review to double check.

Just to make sure I’m not missing anything, is the “native solution” what is described by this [comment](https://github.com/mattermost/mattermost-server/issues/9432#issuecomment-541347755)? If so, I could try implementing this too. Of note, this solution removes the smartdash behavior altogether.

Based on whichever solution, I can then cover the inline markdown functionality mentioned as well.

enahum commented:
@ami9000 thanks for looking into this.  if there are no performance implications I'm fine with it... and yes QA will certainly make a pass and determine if it is working as intended and without perf degradation.

So lets wait for the QA pass and in case there is perf degradation or changing the keyboard type is not an approve UX, then we can consider implementing the native behavior.

If we end up implementing it natively, perhaps we can add a function to the module so we can enable/disable at will. just a thought.

chuttam commented on app/components/post_textbox/post_textbox_base.js:
```diff
@@ -407,17 +408,27 @@ export default class PostTextBoxBase extends PureComponent {
 
-    handlePostDraftSelectionChanged = (event) => {
-        const cursorPosition = event.nativeEvent.selection.end;
+    handleOnSelectionChange = (e) => {
```
Rename `e` to `event` to align with the other function definitions.... easier to search in the codebase too.

mattermod commented:
This issue has been automatically labelled "stale" because it hasn't had recent activity.
A core team member will check in on the status of the PR to help with questions.
Thank you for your contribution!

/cc @jasonblais @hanzei

jasonblais commented:
@migbot or @enahum: quick reminder to help with reviews :)

ami9000 commented on app/components/post_textbox/post_textbox_base.js:
```diff
@@ -407,17 +408,27 @@ export default class PostTextBoxBase extends PureComponent {
-    handlePostDraftSelectionChanged = (event) => {
-        const cursorPosition = event.nativeEvent.selection.end;
+    handleOnSelectionChange = (event) => {
+        if (event) {
```
Yes. If this question is towards the function as a whole: This function was created to help determine whether the `handlePostDraftSelectionChanged` call came from the textInput's `onChangeText` or `onSelectionChange`. Here, `handlePostDraftSelectionChanged` gets called with second argument as `false` (and `true` with `onChangeText`). `onChangeText` doesn't get passed with event/cursorPosition data so `handlePostDraftSelectionChanged` will know via this second argument to pull cursorPosition from state instead.

If this was just about why there's an `if (event)` - just a double check. Not needed though.
